### PR TITLE
Table partitioning and window partitioning must match.

### PR DESCRIPTION
**Pull request by BillWhite**
Table partitioning and window partitioning are related.  If a query has
a windowed aggregate with a PARTTION BY expression which is a reference
to a table partition column, then the PartitionByPlanNode from the
windowed aggregate expression can be pushed down to a distributed
fragment.  Otherwise, the PartionByPlanNode needs to be executed on the
coordinator fragment.

This commit implements this properly.

### CODE REVIEW

File: src/frontend/org/voltdb/planner/parseinfo/StmtSubqueryScan.java [hunk 1]
```diff
@@ -348,6 +348,9 @@ public boolean canRunInOneFragment() {
                 return false;
             }
         }
+        if ( ! selectStmt.hasPartitionColumnInWindowedExpression()) {
+            return false;
```
Comment (wolffcm) [inline src/frontend/org/voltdb/planner/parseinfo/StmtSubqueryScan.java]: Not sure I understand this, because you could have
```SQL
select * from (select rank() over (partition by pkcol order by v) from t where pkcol = ?) as t1;
```
That would be a single-partition statement, so could be just one fragment.  It reasonably matches the logic above for GROUP BY though, so it seems like it's probably okay.  There's probably some context I'm missing?

File: tests/frontend/org/voltdb/regressionsuites/TestWindowedAggregateSuite.java [hunk 2]
```diff
@@ -479,6 +488,47 @@ public void testRankOrderbyExpressions() throws Exception {
...
+
+    private void validateSubqueryWithWindowedAggregate(Client client, String sql)
+            throws IOException, NoConnectionsException, ProcCallException {
+        ClientResponse cr;
+        VoltTable vt;
+        cr = client.callProcedure("@AdHoc", sql);
+        assertEquals(ClientResponse.SUCCESS, cr.getStatus());
+        vt = cr.getResults()[0];
+        int nc = vt.getColumnCount();
+        while (vt.advanceRow()) {
```
Comment (wolffcm) [inline tests/frontend/org/voltdb/regressionsuites/TestWindowedAggregateSuite.java]: You are using the outer query to verify the results of the inner query.  Interesting.

**Comment by wolffcm** [comment]
This looks good, especially after your verbal explanation.

File: tests/frontend/org/voltdb/regressionsuites/TestWindowedAggregateSuite.java (hunk 2 above)
Comment (BillWhite) [inline tests/frontend/org/voltdb/regressionsuites/TestWindowedAggregateSuite.java]: This is really John Crawford's idea.

File: src/frontend/org/voltdb/planner/ParsedSelectStmt.java [hunk 3]
```diff
@@ -825,7 +826,7 @@ private void verifyWindowedExpressions() {
         // Check for windowed expressions.
         if (m_windowedExpressions.size() > 0) {
             if (m_windowedExpressions.size() > 1) {
-                throw new PlanningErrorException("At most one windowed display column is supported.");
+                throw new PlanningErrorException("Only one windowed RANK() expression may appear in a display list.");
```
Comment (paulmartel) [inline src/frontend/org/voltdb/planner/ParsedSelectStmt.java]: consistency: Is it a "select list" as in the message changed above or a "display list"? I suggest consulting Andrew or the doc.

File: src/frontend/org/voltdb/planner/parseinfo/StmtSubqueryScan.java (hunk 1 above)
Comment (paulmartel) [inline src/frontend/org/voltdb/planner/parseinfo/StmtSubqueryScan.java]: @wolffcm. Yeah, some of the call structure and nomenclature in this area is a little hard to follow. I think that the context of this call is the analysis of a KNOWN MP plan tree to decide whether it contains any significant processing on the coordinator that would cause incorrect results if the plan were executed in its entirety on the partition as part of a more complicated distributed fragment.

File: tests/frontend/org/voltdb/planner/TestWindowedFunctions.java [hunk 4]
```diff
@@ -214,6 +214,52 @@ public void testRankWithSubqueries() {
         validateQueryWithSubquery(windowedQuery);
         windowedQuery = "SELECT BBB.B, RANK() OVER (PARTITION BY ALPHA.A ORDER BY BBB.B ) AS ARANK FROM (select A, B, C from AAA where A < B) ALPHA, BBB WHERE ALPHA.C <> BBB.C;";
         validateQueryWithSubquery(windowedQuery);
+
+        // Test with windowed aggregates in the subquery itself.
+
+        // First, use a window partition by which is a table partition column. The partition by node can be
```
Comment (paulmartel) [inline tests/frontend/org/voltdb/planner/TestWindowedFunctions.java]: "use a window partition by which is a table partition column" would be much easier to parse if you quoted or upper cased PARTITION BY so that it was more clearly a noun phrase.

File: tests/frontend/org/voltdb/planner/TestWindowedFunctions.java [hunk 5]
```diff
@@ -214,6 +214,52 @@ public void testRankWithSubqueries() {
...
+        windowedQuery = "SELECT * FROM ( SELECT A, B, C, RANK() OVER (PARTITION BY A ORDER BY B) FROM AAA_PA) ARANK;";
+        validateQueryWithSubqueryWithWindowedAggregate(windowedQuery, 0, 1);
+
+        // Now, use a window partition by which is not in the table partition column.  The partition by
+        // node can no longer be distributed.  So we expect it to show up in the coordinator fragment.
+        windowedQuery = "SELECT * FROM ( SELECT A, B, C, RANK() OVER (PARTITION BY B ORDER BY A) FROM AAA_PA ) ARANK;";
+        validateQueryWithSubqueryWithWindowedAggregate(windowedQuery, 1, 0);
+
+        // Test that putting a windowed aggregate in the outer select list gets about the
+        // same answers.  The outer windowed aggreate adds 1 to all the PB counts on the
```
Comment (paulmartel) [inline tests/frontend/org/voltdb/planner/TestWindowedFunctions.java]: spelling: aggreate

File: tests/frontend/org/voltdb/regressionsuites/TestWindowedAggregateSuite.java [hunk 6]
```diff
@@ -494,6 +544,10 @@ public void testExplainPlan() throws Exception {
         try {
             cr = client.callProcedure("@Explain", sql);
             assertEquals(ClientResponse.SUCCESS, cr.getStatus());
+            // We just care that the explain has succeeded and not
+            // caused an NPE.  The results here are not used, but they
+            // are useful for diagnosing errors.  So we leave them
+            // here.
             VoltTable vt = cr.getResults()[0];
             assertTrue(true);
```
Comment (paulmartel) [inline tests/frontend/org/voltdb/regressionsuites/TestWindowedAggregateSuite.java]: Nice comment. Yet `assertTrue(true);` is STILL an affront to meaningful coding.

**Comment by paulmartel** [comment]
That's all from me.
